### PR TITLE
Add Data List Directive with ngdocs and unit tests

### PR DESCRIPTION
**Pull request by jeff-phillips-18**
Also adds a utility transclude directive to allow isolated scopes on
items inside a directive.

### CODE REVIEW

File: src/utils/transclude-directive.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
+
```
Comment (waldenraines) [inline src/utils/transclude-directive.js]: This is lacking documentation.

File: src/views/datalist/data-list-directive.js [hunk 2]
```diff
@@ -0,0 +1,312 @@
+/**
+ * @ngdoc directive
+ * @name patternfly.views.directive:pfDataList
+ *
+ * @description
+ *   Directive for rendering a data list.
+ *   <br><br>
+ *
+ * @param {object} config configuration settings for the sparkline chart:<br/>
```
Comment (dtaylor113) [inline src/views/datalist/data-list-directive.js]: cut & paste error "...settings for the sparkline chart"

**Comment by dtaylor113** [comment]
Hi, I just cloned your repo, npm install, bower install, grunt build, grunt ngdocs:view and when I click on the pfDataList ngdoc, I get:

angular.module('patternfly.views').controller('ViewCtrl', ['$scope',
  function ($scope) {
    $scope.eventText = '';
    var handleSelect = function (item, e) {
...
...
 SyntaxError: Unexpected token ;
    at eval (native)
    at http://localhost:8000/js/angular-bootstrap-prettify.js:179:20
    at Object.forEach (http://localhost:8000/grunt-scripts/angular.js:326:20)
    at Object.compile (http://localhost:8000/js/angular-bootstrap-prettify.js:173:15)
    at applyDirectivesToNode (http://localhost:8000/grunt-scripts/angular.js:7530:32)
...
...
Error: [ng:areq] Argument 'ViewCtrl' is not a function, got undefined
http://errors.angularjs.org/1.3.18/ng/areq?p0=ViewCtrl&p1=not%20a%20function%2C%20got%20undefined
    at REGEX_STRING_REGEXP (angular.js:63)
    at assertArg (angular.js:1590)
    at assertArgFn (angular.js:1600)

Not sure if this is my enviornment or something in the repo?

~/Dev/repos/j-a-pf>bower list
bower check-new     Checking for new versions of the project dependencies...
angular-patternfly#2.0.0 /home/dtaylor/Dev/repos/j-a-pf
├── angular#1.3.18 (latest is 1.4.4)
├─┬ angular-animate#1.3.18 (latest is 1.4.4)
│ └── angular#1.3.18 (latest is 1.4.4)
├─┬ angular-mocks#1.3.18 (latest is 1.4.4)
│ └── angular#1.3.18
├─┬ angular-route#1.3.18 (latest is 1.4.4)
│ └── angular#1.3.18
├─┬ angular-sanitize#1.3.18 (latest is 1.4.4)
│ └── angular#1.3.18
├─┬ angular-touch#1.3.18 (latest is 1.4.4)
│ └── angular#1.3.18
├── lodash#3.10.1
└── patternfly#2.0.0

File: src/views/datalist/data-list-directive.js [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+  function () {
+    'use strict';
+    return {
+      restrict: 'A',
+      scope: {
+        config: '=',
+        items: '=',
+        eventId: '@id'
+      },
+      replace: true,
```
Comment (waldenraines) [inline src/views/datalist/data-list-directive.js]: This is [deprecated](https://docs.angularjs.org/api/ng/service/$compile#-replace-deprecated-will-be-removed-in-next-major-release-i-e-v2-0-).

File: src/views/datalist/data-list-directive.js [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+            showSelectBox: true,
+            rowHeight: 36,
+            onSelect: null,
+            onSelectionChange: null,
+            onCheckBoxChange: null,
+            onClick: null,
+            onDblClick: null
+          };
+
+          $scope.config = $.extend(true, angular.copy($scope.defaultConfig), $scope.config);
```
Comment (waldenraines) [inline src/views/datalist/data-list-directive.js]: Why is this needed here if it's done again on `attrs.$observe()`?

File: styles/angular-patternfly.css [hunk 5]
```diff
@@ -79,3 +79,75 @@
...
+  margin-top: 0;
+}
+.data-list-pf .list-group-item .list-column {
+  float: left;
+  padding: 5px 0px;
+}
+.data-list-pf .list-group-item .pficon {
+  -webkit-align-items: center;
+  align-items: center;
+  color: #1186C1 !important;
```
Comment (waldenraines) [inline styles/angular-patternfly.css]: Why is the `!important` needed?  Can we instead override with selector specificity?

File: styles/angular-patternfly.css [hunk 6]
```diff
@@ -79,3 +79,75 @@
...
+.data-list-pf .list-group-item:hover,
+.data-list-pf .list-group-item:focus {
+  background-color: #ededed;
+  border-color: #ededed;
+}
+
+.list-group-item.active .pficon,
+.list-group-item.active:hover .pficon,
+.list-group-item.active:focus .pficon {
+  color: #ffffff !important;
```
Comment (waldenraines) [inline styles/angular-patternfly.css]: Why is the `!important` needed?  Can we instead override with selector specificity?

**Comment by jeff-phillips-18** [comment]
Dave,  sorry about that. It was a merge issue when I rebased with Angular 1.3. Fix is coming soon.

File: src/views/datalist/data-list-directive.js (hunk 4 above)
Comment (jeff-phillips-18) [inline src/views/datalist/data-list-directive.js]: This is done to setup the default for cases where config is not set. I should have put =? on config so that it is optional for those who want pure default behavior.

File: src/utils/transclude-directive.js (hunk 1 above)
Comment (jeff-phillips-18) [inline src/utils/transclude-directive.js]: Added ngdoc. I'm not sure unit tests are applicable here.

File: src/views/datalist/data-list-directive.js (hunk 3 above)
Comment (jeff-phillips-18) [inline src/views/datalist/data-list-directive.js]: Thanks, removed it.

File: styles/angular-patternfly.css (hunk 5 above)
Comment (jeff-phillips-18) [inline styles/angular-patternfly.css]: It's not needed. I've removed it.

File: styles/angular-patternfly.css (hunk 6 above)
Comment (jeff-phillips-18) [inline styles/angular-patternfly.css]: Added the data-list-pf selector and removed !important

**Comment by dtaylor113** [comment]
Ok, thanks, LGTM

File: src/utils/transclude-directive.js (hunk 1 above)
Comment (waldenraines) [inline src/utils/transclude-directive.js]: Why wouldn't they be?

Comment (jeff-phillips-18) [inline src/utils/transclude-directive.js]: Just not entirely sure how to test this.

File: src/utils/transclude-directive.js [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+              // Remove the original directive
+              $delegate.shift();
+              return $delegate;
+          }]);
+      })
+
+      .directive( 'transcludeSibling', function() {
+        return {
+          restrict: 'E',
+          transclude: true,
```
Comment (erundle) [inline src/utils/transclude-directive.js]: Any reason why these are restricted to just 'E'? normal transcludes support both

Comment (jeff-phillips-18) [inline src/utils/transclude-directive.js]: No reason. Just only necessary here based on how the example is written.

File: src/utils/transclude-directive.js (hunk 1 above)
Comment (waldenraines) [inline src/utils/transclude-directive.js]: You'd just test each type of transclusion for the correct scope and DOM structure.  Should be pretty similar to other directive tests.

Comment (jeff-phillips-18) [inline src/utils/transclude-directive.js]: I'm struggling on trying to create directives on the fly in the test spec which is what I need to do in order to test the scope and DOM structures based on the transclusion type. The example in ng-doc show this and tests this pretty well.

I'd like to defer this. We can add an issue and I will address when I can. I need to get a 2.1 label on angulary-patternfly for today as there are folks that need the latest today and I'd like for the data list to be included.

File: Gruntfile.js [hunk 8]
```diff
@@ -116,6 +116,7 @@ module.exports = function (grunt) {
             'lib/patternfly/components/d3/d3.js',
             'lib/angular/angular.js',
             'lib/angular-animate/angular-animate.js',
+            'lib/lodash/lodash.min.js',
```
Comment (erundle) [inline Gruntfile.js]: I don't love the addition of lodash here even though I am supportive of using the library within applications. We need to make sure that future commits don't use functions that are in lodash but not in underscore since these need to be interchangeable.

File: dist/angular-patternfly.js [hunk 9]
```diff
@@ -1978,6 +1990,176 @@ angular.module('patternfly.select', []).directive('pfSelect', function ($timeout
...
+ * <ul style='list-style-type: none'>
+ * <li> 'sibling' - The transcluded contents scope is a sibling one to the element where transclusion happens (default)
+ * <li> 'parent'  - The transcluded contents scope is that of the element where transclusion happens.
+ * <li> 'child'   - The transcluded contents scope is child scope to the scope of the element where transclusion happens.
+ * </ul>
+ *
+ * @description
+ *   Directive for transcluding in directives and setting up scope of children of parent directives. This is a workaround
+ *   for https://github.com/angular/angular.js/issues/5489
+ *
```
Comment (erundle) [inline dist/angular-patternfly.js]: Is there any reason why you can't use use compile and add the secondary directive to the DOM after the first one has been processed. I've done something like that in the past but I don't know your use case.

File: dist/angular-patternfly.js [hunk 10]
```diff
@@ -1978,6 +1990,176 @@ angular.module('patternfly.select', []).directive('pfSelect', function ($timeout
...
+        height: 20px;
+        text-align: center;
+        line-height: 20px;
+        margin-left: 5px;
+      }
+
+      .pf-transclude-example pre {
+        padding: 5px;
+        border-width: 0px;
+      }
```
Comment (erundle) [inline dist/angular-patternfly.js]: Future note we need to add a demo.css file that doesn't get distributed via bower. I will create an issue for it

File: Gruntfile.js (hunk 8 above)
Comment (jeff-phillips-18) [inline Gruntfile.js]: Agreed.

File: dist/angular-patternfly.js (hunk 10 above)
Comment (jeff-phillips-18) [inline dist/angular-patternfly.js]: This is only in example code so I don't see it as a huge issue.

Comment (waldenraines) [inline dist/angular-patternfly.js]: Why not just create another `<file>` with the styles in it instead of putting them in the html?

Comment (jeff-phillips-18) [inline dist/angular-patternfly.js]: Done.

File: dist/angular-patternfly.js (hunk 9 above)
Comment (jeff-phillips-18) [inline dist/angular-patternfly.js]: I see this as a utility that will be used time and again as add more complex directives.  This makes it easy on the directives and it's better to have it all in one place in my opinion.

**Comment by jeff-phillips-18** [comment]
Squashed and rebased.

**Comment by jeff-phillips-18** [comment]
@waldenraines @erundle Can this be merged?

**Comment by erundle** [comment]
I am fine with approach pending you do investigate writing additional unit test later today/monday. They have proven very useful when upgrading components etc and is something that is need. We have already had a request to support 1.4.x

**Comment by waldenraines** [comment]
> @waldenraines @erundle Can this be merged?

Can we create a GH issue for adding the unit tests?
